### PR TITLE
2512: Allow more than 5 tabs

### LINKED ISSUE #2512: More than five tabs
Please can we have more than five tabs - either by squashing them together a bit more or by allowing to scroll the tabs bar left/right to see the ones which are out of view?

Home, Mentions, Local, Federated, Direct Messages is already five tabs before adding ones for lists or hashtags.

### PR DESCRIPTION
**Pull request by Lakoja**
Fixes #2512 

grafik

Can add an arbitrary number of tabs.
Graphical behavior is unchanged for small numbers: the whole space if filled with the tabs - they are enlarged if needed.

If there are more the mode switches to "scrollable".
This does not, however, look very differently (see screenshot with the current tab scrolled out).

### CODE REVIEW

**Comment by connyduck** [comment]
Why not just use `app:tabMode="auto"` instead of a custom view?

**Comment by Lakoja** [comment]
> Why not just use `app:tabMode="auto"` instead of a custom view?

Yes, I tried that first.
It didn't work for me. (And there are a few SO topics with that content as well.)

**Comment by connyduck** [comment]
What does "didn't work" mean exactly? I just tried it, seems fine.

**Comment by Lakoja** [comment]
> What does "didn't work" mean exactly? I just tried it, seems fine.

Maybe a version issue?

It behaved exactly like "scrollable": No enlargement for few items.

**Comment by Lakoja** [comment]
This is how it looks with "auto" on Android 11 (emulator):

grafik

**Comment by connyduck** [comment]
Ok it is not stretching the items when they are few. Seems like it is a known bug since 4 years but they don't want to fix it because it is a breaking api change. https://github.com/material-components/material-components-android/issues/650
oh well

File: app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * Idea taken from https://stackoverflow.com/a/44894143
+ */
+class AdaptiveTabLayout : TabLayout {
+    constructor(context: Context) : super(context)
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
+        context,
+        attrs,
+        defStyleAttr
+    )
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt:21]:
```suggestion
class AdaptiveTabLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : TabLayout(context, attrs, defStyleAttr) {
```

File: app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package com.keylesspalace.tusky.util
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt:1]: should be in the `com.keylesspalace.tusky.view` package

File: app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                widthOfAllTabs += tabLayout.getChildAt(i).measuredWidth
+            }
+            if (widthOfAllTabs <= measuredWidth) {
+                // fill all space if there is enough room
+                tabMode = MODE_FIXED
+                super.onMeasure(widthMeasureSpec, heightMeasureSpec)
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt:44]: Catching all exceptions is rarely a good idea - and I don't see what would even throw here?

(also exceptions should be logged with `android.util.Log` and not `printStackTrace()`)

**Review by connyduck — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt (hunk 3 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt]: Removed (was copy&paste)

File: app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt (hunk 2 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/util/AdaptiveTabLayout.kt]: ok

**Review by connyduck — APPROVED** [review]

**Comment by Lakoja** [comment]
> My fault for not trying until after I merged, but it looks like this for me (pixel 2xl android 11)
> 
> Screenshot_20231011-100700

So, all the colors are wrong?

Especially also white is not correct?

You can easily disable this change: Revert `activity_main.xml`

Is it correct then?

**Comment by Lakoja** [comment]
> My fault for not trying until after I merged, but it looks like this for me (pixel 2xl android 11)
> 
> Screenshot_20231011-100700

https://github.com/tuskyapp/Tusky/pull/4058

Trying to fix it there.

However, I have no idea why this happens (now) and also my fix does not fully work.

(You can set the tab icon color, but not when it is selected.)

**Pull request merged**